BYO LLVM build configuration broken by pybind11 changes

See https://github.com/iree-org/iree/pull/18872#issuecomment-2435492981

Logs: https://github.com/iree-org/iree/actions/runs/11496514385/job/31998278068#step:4:10340

```
-- Found Python3: /__w/iree/iree/build-byo-llvm/.venv/bin/python3 (found version "3.11.0") found components: Interpreter Development Development.Module Development.Embed 
-- Found Python3: /__w/iree/iree/build-byo-llvm/.venv/bin/python3 (found version "3.11.0") found components: Interpreter Development.Module 
-- Found python include dirs: /usr/include/python3.11
-- Found python libraries: /usr/lib/x86_64-linux-gnu/libpython3.11.so
-- Found numpy v: 
-- Checking for pybind11 in python path...
-- not found (install via 'pip install pybind11' or set pybind11_DIR)
CMake Error at cmake/modules/MLIRDetectPythonEnv.cmake:28 (find_package):
  Could not find a package configuration file provided by "pybind11"
  (requested version 2.9) with any of the following names:

    pybind11Config.cmake
    pybind11-config.cmake

  Add the installation prefix of "pybind11" to CMAKE_PREFIX_PATH or set
  "pybind11_DIR" to a directory containing one of the above files.  If
  "pybind11" provides a separate development package or SDK, be sure it has
  been installed.
Call Stack (most recent call first):
  CMakeLists.txt:175 (mlir_configure_python_dev_packages)


-- Configuring incomplete, errors occurred!
```

Code:

* https://github.com/iree-org/iree/blob/main/.github/workflows/ci_linux_x64_clang_byollvm.yml
* https://github.com/iree-org/iree/blob/main/build_tools/cmake/build_and_test_byo_llvm.sh
* https://github.com/iree-org/iree/blob/main/build_tools/llvm/byo_llvm.sh
* https://github.com/iree-org/iree/blob/main/build_tools/cmake/iree_llvm.cmake

---

We can also increase the frequency of https://github.com/iree-org/iree/blob/main/.github/workflows/ci_linux_x64_clang_byollvm.yml (from nightly to every commit) if we have CI budget/capacity, and add alerting (like posting to Discord on failure).

--- a/build_tools/cmake/build_and_test_byo_llvm.sh
+++ b/build_tools/cmake/build_and_test_byo_llvm.sh
@@ -24,6 +24,12 @@ echo "Setting up venv at $VENV_DIR"
 python3 -m venv "$VENV_DIR"
 source "$VENV_DIR/bin/activate"
 python -m pip install -r runtime/bindings/python/iree/runtime/build_requirements.txt
+python -m pip install -r third_party/llvm-project/mlir/python/requirements.txt
+# Note: IREE's Python bindings for Python 3.13 are build with support for
+# free-threading for which support was added to pybind with version 2.13.0.
+# Therefore, we upgrade to a more recent version and avoid mixing of different
+# pybind versions.
+python -m pip install pybind11==2.13.6
 
 # Note: by using the `build_llvm` action here, we are exercising byo_llvm.sh's
 # ability to build LLVM... from our own third_party/llvm-project. That's not

--- a/build_tools/llvm/byo_llvm.sh
+++ b/build_tools/llvm/byo_llvm.sh
@@ -113,6 +113,9 @@ do_build_mlir() {
 
   cmake_options="-DLLVM_DIR='${main_install_dir}/lib/cmake/llvm'"
   cmake_options="${cmake_options} -DPython3_EXECUTABLE='$(which $python3_command)'"
+  # Note: Building the MLIR Python bindings requires the installation of
+  # dependencies as specified in `mlir/python/requirements.txt`, which among
+  # others include pybind11.
   cmake_options="${cmake_options} -DMLIR_ENABLE_BINDINGS_PYTHON=ON"
   cmake_options="${cmake_options} -DCMAKE_INSTALL_PREFIX=${mlir_install_dir}"
   cmake_options="${cmake_options} -C $TD/mlir_config.cmake"

